### ISSUE TITLE
Remove experimentalRetryJoin

### ISSUE THREAD

**Issue by killianmuldoon** [kubernetes-sigs/cluster-api#6942]
experimentalRetryJoin is [deprecated as of CAPI release 1.2](https://github.com/kubernetes-sigs/cluster-api/pull/6938). Removal is expected to happen in release 1.3 - but before proceeding we should have high confidence that CAPI + Kubeadm are capable today of solving [the problem that experimentalRetryJoin was designed to fix](https://github.com/kubernetes-sigs/cluster-api/pull/2763/files).

In order to increase confidence in the removal we should:
1. Ask providers which, if any, are still using experimentalRetryJoin, to what extent, and to solve what problem.
2. Figure out a testing strategy for providers that would trigger the problem experimentalRetryJoin was designed to fix.
3. Ensure these tests are passing and we have high confidence in the current retryJoin behaviour of Kubeadm.

Once we're confident that we can proceed with removal we need to: 
1. Remove the field from the bootstrap KubeadmConfig types + regenerate CRDs
2. Stop embedding the kubeadm-bootsrap-script in `bootstrap/kubeadm/internal/cloudinit/cloudinit.go`
3. Update cloudinit.go, removing functions and variables related to the embedded script e.g. generateBootstrapScript, kubeadmBootstrapScript
4. Remove the KubeadmBootstrap script
5. Add a note to the migration guide that the 

/area control-plane

**Comment by sbueringer** [comment]
In my opinion we should never remove a field from a v1beta1 apiVersion without an apiVersion bump.

**Comment by killianmuldoon** [comment]
> In my opinion we should never remove a field from a v1beta1 apiVersion without an apiVersion bump as it's a breaking change to v1beta1 otherwise.
 
Good point, but I suppose that doesn't necessarily preclude us from changing the behaviour. It's not exactly a feature gate, but the field is prefixed with `experimental`

**Comment by sbueringer** [comment]
So we would keep the field but disable its behavior? I think that's a very dangerous precedent.

**Comment by killianmuldoon** [comment]
> So we would keep the field but disable its behavior? I think that's a very dangerous precedent and ~ the same as dropping the field.

This is done regularly for experimental features as I understand it. The key question is whether we consider this experimental or not given the prefix. I think the most important thing here is whether the actual behaviour is fixed rather than the API fields and behaviour though.

**Comment by vincepri** [comment]
/kind api-change

As mentioned above, we need to deprecate now (just a comment on the field would suffice) + remove in the next API version

**Comment by fabriziopandini** [comment]
/milestone v1.3

to track to to this before the release

**Comment by sbueringer** [comment]
If I'm not mistaken deprecation was done in these PRs:
* https://github.com/kubernetes-sigs/cluster-api/pull/6938
* https://github.com/kubernetes-sigs/cluster-api/pull/6941

It was already part of v1.2.0 (https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.0).

I wonder a bit about removal though.

I don't have background information for what exactly this feature was needed and why we can drop it now (either because it's not needed anymore or we have a replacement I guess (?)).

Additional indication for that:
> @yastij: why did we drop UseExperimentalRetryJoin it's still needed in slow/flaky envs
> https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1602#discussion_r965996706

**Comment by killianmuldoon** [comment]
I think we should remove the good first issue and help wanted labels, as this removal isn't something easily implemented. @fabriziopandini WDYT?

**Comment by fabriziopandini** [comment]
I I got this right what we are doing now is just applying a deprecated label on the API field (I have updated the issue title)

**Comment by sbueringer** [comment]
> If I'm not mistaken deprecation was done in these PRs:
> 
> * [seedling  Deprecate experimentalRetryJoin #6938](https://github.com/kubernetes-sigs/cluster-api/pull/6938)
> * [[release-1.2] seedling  Deprecate experimentalRetryJoin #6941](https://github.com/kubernetes-sigs/cluster-api/pull/6941)
> 
> It was already part of v1.2.0 (https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.0).
> 
> I wonder a bit about removal though.
> 
> I don't have background information for what exactly this feature was needed and why we can drop it now (either because it's not needed anymore or we have a replacement I guess (?)).
> 
> Additional indication for that:
> 
> > @yastij: why did we drop UseExperimentalRetryJoin it's still needed in slow/flaky envs
> > [kubernetes-sigs/cluster-api-provider-vsphere#1602 (comment)](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1602#discussion_r965996706)

I'm slightly confused. As mentioned, I think we did this already?

**Comment by fabriziopandini** [comment]
Dropping to the milestone
@yastij @randomvariable what is your opinion about making this happen in 1.4 timeframe (target date ~march 2023)

**Comment by sbueringer** [comment]
Just a note. Just checked linked issues and talked to some folks. We should be really able to remove the field as there were various improvements to kubeadm.

Also in the worst case were folks still need it they can remove the kubeadm binary through a script and implement their own retry handling there.

**Comment by neolit123** [comment]
> Just a note. Just checked linked issues and talked to some folks. We should be really able to remove the field as there were various improvements to kubeadm.
> 
> Also in the worst case were folks still need it they can remove the kubeadm binary through a script and implement their own retry handling there.

+1, i though this was removed already. :\

**Comment by neolit123** [comment]
i see there are steps in the OP, i can give this a shot.
/assign

**Comment by sbueringer** [comment]
We have to wait for the next apiVersion

**Comment by fabriziopandini** [comment]
based on the comment above
/triage accepted

**Comment by chrischdi** [comment]
Note: https://github.com/kubernetes-sigs/cluster-api/pull/10983 gives another indicator that we should not re-add this in v1beta2.
